keps/sig-network/3866-nftables-proxy/README.md

```diff
@@ -0,0 +1,1538 @@
+  adopting upstream.
+
+[kube-nftlb]: https://github.com/zevenet/kube-nftlb
+[nftlb]: https://github.com/zevenet/nftlb
+[nfproxy]: https://github.com/sbezverk/nfproxy
+[kpng's nft backend]: https://github.com/kubernetes-sigs/kpng/tree/master/backends/nft
+
+### Risks and Mitigations
+
+The primary risk of the proposal is feature regressions, which will be
+addressed by testing, and by a slow, optional, rollout of the new proxy
+mode.
+
+The `nftables` mode should not pose any new security issues relative
+to the `iptables` mode.
+
+## Design Details
+
+### High level design
+
+At a high level, the new mode should have the same architecture as the
+existing modes; it will use the service/endpoint-tracking code in
+`k8s.io/kubernetes/pkg/proxy` to watch for changes, and update rules
+in the kernel accordingly.
+
+### Low level design
+
+Some details will be figured out as we implement it. We may start with
+an implementation that is architecturally closer to the `iptables`
+mode, and then rewrite it to take advantage of additional nftables
+features over time.
+
+#### Tables
+
+Unlike iptables, nftables does not have any reserved/default tables or
+chains (eg, `nat`, `PREROUTING`). Users are expected to create their
+own tables and chains for their own purposes. An nftables table can
+only contain rules for a single "family" (`ip` (v4), `ip6`, `inet`
+(both IPv4 and IPv6), `arp`, `bridge`, or `netdev`), but unlike in
+iptables, you can have both "filter"-type chains and "NAT"-type chains
+in the same table.
+
+So, we will create a single `kube_proxy` table in the `ip` family, and
+another in the `ip6` family. All of our chains, sets, maps, etc, will
+go into those tables. Other system components (eg, firewalld) should
+ignore our table, so we should not need to worry about watching for
+other people deleting our rules like we have to in the `iptables`
+backend.
+
+(In theory, instead of creating one table each in the `ip` and `ip6`
+families, we could create a single table in the `inet` family and put
+both IPv4 and IPv6 chains/rules there. However, this wouldn't really
+result in much simplification, because we would still need separate
+sets/maps to match IPv4 addresses and IPv6 addresses. (There is no
+data type that can store/match either an IPv4 address or an IPv6
+address.) Furthermore, because of how Kubernetes Services evolved in
+parallel with the existing kube-proxy implementation, we have ended up
+with a dual-stack Service semantics that is most easily implemented by
+handling IPv4 and IPv6 completely separately anyway.)
+
+#### Communicating with the kernel nftables subsystem
+
+At least initially, we will use the `nft` command-line tool to read
+and write rules, much like how we use command-line tools in the
+`iptables` and `ipvs` backends. However, the `nft` tool is mostly just
+a thin wrapper around `libnftables`, and it would be possible to use
+that directly instead in the future, given a cgo wrapper.
```

Comment:
Huh. I didn't realize we produced binaries that people ran outside of containers that we also produced.

Anyway, we don't _have to_ use the library. I was mostly pointing out the "thin wrapper" part and suggesting that even if our golang API has obvious nft-command-line-isms in it, that wouldn't prevent us from switching to using libnftables later.

Reply:
We compile almost everything with CGO_ENABLED=0 to simplify cross compiling and distribution but not kubelet and I think we get by there mostly because the build env is Debian stable so most distros have >= libc etc.

Fair, we also don't have to not use it, but non-static binaries do make build & distribution trickier.